lib/phoenix/tracker/state.ex

```diff
@@ -312,22 +318,77 @@ defmodule Phoenix.Tracker.State do
     {new_state, joins, leaves}
   end
 
-  @spec accumulate_joins(t, values) :: joins :: {[pid_lookup], [values]}
+  @spec accumulate_joins(t, values) :: joins :: {[pid_lookup], [values], [tag_lookup]}
   defp accumulate_joins(local, remote_map) do
     %State{context: context, clouds: clouds} = local
-    Enum.reduce(remote_map, {[], []}, fn {{replica, _} = tag, {pid, topic, key, meta}}, {pids, adds} ->
+    Enum.reduce(remote_map, {[], [], []}, fn {{replica, _} = tag, {pid, topic, key, meta}}, {pids, adds, tags} ->
       if not match?(%{^replica => _}, context) or in?(context, clouds, tag) do
-        {pids, adds}
+        {pids, adds, tags}
       else
-        {[{pid, topic, key} | pids], [{{topic, pid, key}, meta, tag} | adds]}
+        {
+          [{pid, topic, key} | pids],
+          [{{topic, pid, key}, meta, tag} | adds],
+          [{tag, {topic, pid, key}} | tags]
+        }
       end
     end)
   end
 
+  # This method is used only when remote context is empty (which it is when
+  # merging local state and a remote delta). This is an optimized method that
+  # uses pid lookup ets table to fetch and delete only values that need to be
+  # deleted.
   @spec observe_removes(t, t, map) :: {clouds, delta, leaves :: [value]}
-  defp observe_removes(%State{pids: pids, values: values, delta: delta} = local, remote, remote_map) do
+  defp observe_removes(
+    %State{pids: pids, tags: tags, values: values, delta: delta} = local,
+    %State{context: remote_context, clouds: remote_clouds} = remote,
+    remote_map
+  ) when remote_context == %{} do
+    unioned_clouds = union_clouds(local, remote)
+    init = {unioned_clouds, delta, []}
+
+    # Remote clouds show which tags were modified. If the tag is not in the
+    # remote map then that tag was removed.
+    tags_to_remove = Enum.flat_map(remote_clouds, fn {replica, cloud} ->
+      if replica == local.replica do
+        []
+      else
+        Enum.reject(cloud, &tag_in_remote_map?(&1, remote_map))
+      end
+    end)
+
+    Enum.reduce(tags_to_remove, init, fn tag, {clouds, delta, leaves} ->
+      case :ets.lookup(tags, tag) do
+        [{_tag, {topic, pid, key} = values_key}] ->
+          [el] = :ets.lookup(values, values_key)
+
+          :ets.delete(values, values_key)
```

Comment:
These 4 lines are repeated in the other function case, capture in a function.

Reply:
https://github.com/salemove/phoenix_pubsub/compare/dce3bc868efd51d59f320102402a7ac040cb38aa..a27ea2e71f2a7b154e178d6a82bf3f93b7536586

Reply:
I think the next line can also be captured there.

Reply:
I initially wanted but that would mean adding 3 more parameters to the function which makes it quite ugly and then that function cannot be reused in remove_down_replicas function.